server/lib/buff_server/packages.ex

```diff
@@ -26,80 +24,67 @@ defmodule BuffServer.Packages do
   Gets a single package.
 
   Raises `Ecto.NoResultsError` if the Package does not exist.
-
-  ## Examples
-
-      iex> get_package!(123)
-      %Package{}
-
-      iex> get_package!(456)
-      ** (Ecto.NoResultsError)
-
   """
   def get_package!(id), do: Repo.get!(Package, id)
 
   @doc """
-  Creates a package.
-
-  ## Examples
-
-      iex> create_package(%{field: value})
-      {:ok, %Package{}}
-
-      iex> create_package(%{field: bad_value})
-      {:error, %Ecto.Changeset{}}
-
+  Creates the package and uploads to S3, this is run inside a transaction so if anything fails it'll rollback.
   """
-  def create_package(attrs \\ %{}) do
-    %Package{}
-    |> Package.changeset(attrs)
-    |> Repo.insert()
+  def create_package(%{artifact_binary: artifact_binary} = attrs)
+      when is_binary(artifact_binary) do
+    bucket_path = s3_bucket_path(attrs)
+    attrs = Map.merge(attrs, %{s3_bucket_name: @s3_bucket_name, s3_bucket_path: bucket_path})
+    changeset = Package.changeset(%Package{}, attrs)
+
+    transaction_res =
+      Multi.new()
+      |> Multi.insert(:package, changeset)
+      |> Multi.run(:s3_artifact, fn _repo, changes ->
+        put_result =
+          @s3_bucket_name
+          |> ExAws.S3.put_object(bucket_path, artifact_binary)
+          |> ExAws.request()
+
+        with {:ok, %{status_code: 200}} <- put_result do
+          {:ok, changes}
+        else
+          err -> {:error, err}
+        end
+      end)
+      |> Repo.transaction()
+
+    with {:ok, %{package: package}} <- transaction_res do
+      {:ok, package}
+    else
+      err -> err
+    end
   end
 
   @doc """
-  Updates a package.
-
-  ## Examples
-
-      iex> update_package(package, %{field: new_value})
-      {:ok, %Package{}}
-
-      iex> update_package(package, %{field: bad_value})
-      {:error, %Ecto.Changeset{}}
-
+  Generates the s3 bucket path to be used for this package.
   """
-  def update_package(%Package{} = package, attrs) do
-    package
-    |> Package.changeset(attrs)
-    |> Repo.update()
+  def s3_bucket_path(attrs) do
+    "/#{attrs.name}/artifact"
   end
 
   @doc """
   Deletes a Package.
-
-  ## Examples
-
-      iex> delete_package(package)
-      {:ok, %Package{}}
-
-      iex> delete_package(package)
-      {:error, %Ecto.Changeset{}}
-
   """
   def delete_package(%Package{} = package) do
-    Repo.delete(package)
-  end
-
-  @doc """
-  Returns an `%Ecto.Changeset{}` for tracking package changes.
-
-  ## Examples
-
-      iex> change_package(package)
-      %Ecto.Changeset{source: %Package{}}
-
-  """
-  def change_package(%Package{} = package) do
-    Package.changeset(package, %{})
+    Multi.new()
```

Comment:
If we see this pattern repeated more we'll need to think about it, super annoying to write